### PR TITLE
[Feature]Add `PatchEmbed` and `PatchMerging` with `AdaptivePadding`

### PR DESCRIPTION
**Pull request by jshilong**
## Motivation

`PatchEmbed` and `PatchMerging` are basic modules that would be used in many transformer methods, such as PVT and Swin
At the same time, `AdaptivePadding` is an essential feature to deal with the dynamic shape in detection. 

## Modification

- Add `AdaptivePadding` to avoid lost pixel of input when doing the patchembed or patchmerge. 
- Add`PatchEmbed` and `PatchMerging`  to `/mmdet/models/utils/transformer.py` and add the corresponding unitests.
- Always apply adaptive padding first before doing the `PatchEmbed` and `PatchMerging` to deal with the dynamic shape in detection.

## BC-breaking (Optional)
None

### CODE REVIEW

File: mmdet/models/utils/transformer.py [hunk 1]
```diff
@@ -30,12 +30,21 @@
 
 
 class AdaptivePadding(nn.Module):
+    """Applies padding to input (if needed) so that input can get fully covered
```
Comment (Junjun2016) [inline mmdet/models/utils/transformer.py:33]: Can add more description and some examples.

Comment (Junjun2016) [reply mmdet/models/utils/transformer.py:33]: Why do we need AdaptivePadding?

Comment (Junjun2016) [reply mmdet/models/utils/transformer.py:33]: The difference between AdaptivePadding and default conv.

File: mmdet/models/utils/transformer.py [hunk 2]
```diff
@@ -25,6 +29,345 @@
...
+            if self.adap_padding:
+                input_h, input_w = input_size
+                kernel_h, kernel_w = self.adap_padding.kernel_size
+                stride_h, stride_w = self.adap_padding.stride
+                output_h = math.ceil(input_h / stride_h)
+                output_w = math.ceil(input_w / stride_w)
+                pad_h = (
+                    max((output_h - 1) * stride_h +
+                        (kernel_h - 1) * self.adap_padding.dilation[0] + 1 -
+                        input_h, 0))
```
Comment (Junjun2016) [inline mmdet/models/utils/transformer.py:200]:
```suggestion
                pad_h = max((output_h - 1) * stride_h + (kernel_h - 1) * self.adap_padding.dilation[0] + 1 -                         input_h, 0)
```

File: mmdet/models/utils/transformer.py [hunk 3]
```diff
@@ -25,6 +29,345 @@
...
+                pad_w = (
+                    max((output_w - 1) * stride_w +
+                        (kernel_w - 1) * self.adap_padding.dilation[1] + 1 -
+                        input_w, 0))
```
Comment (Junjun2016) [inline mmdet/models/utils/transformer.py:204]:
```suggestion
                pad_w = 
                    max((output_w - 1) * stride_w +
                        (kernel_w - 1) * self.adap_padding.dilation[1] + 1 -
                        input_w, 0)
```

File: mmdet/models/utils/transformer.py [hunk 4]
```diff
@@ -25,6 +29,345 @@
...
+        >>> stride = 16
+        >>> dilation = 1
+        >>> input = torch.rand(1, 1, 15, 17)
+        >>> adap_pad = AdaptivePadding(
+        >>>     kernel_size=kernel_size,
+        >>>     stride=stride,
+        >>>     dilation=dilation,
+        >>>     padding="same")
+        >>> out = adap_pad(input)
+        >>> # padding to divisible by 16
```
Comment (Junjun2016) [inline mmdet/models/utils/transformer.py:57]: Why is 16? Whether 16 is kernel size or stride?
According to the formulation `pad_h = max((output_h - 1) * stride_h + (kernel_h - 1) * self.adap_padding.dilation[0] + 1`, if kernel_size is not equal to stride, the padded shape will not be divisible by kernel size or strice.

**Review by Junjun2016 — APPROVED** [review]
LGTM, good job!

File: mmdet/models/utils/transformer.py (hunk 4 above)
Comment (jshilong) [reply mmdet/models/utils/transformer.py:57]: I will remove this comment

File: mmdet/models/utils/transformer.py [hunk 5]
```diff
@@ -25,6 +29,340 @@
...
+                output_w = math.ceil(input_w / stride_w)
+                pad_h = max((output_h - 1) * stride_h +
+                            (kernel_h - 1) * self.adap_padding.dilation[0] +
+                            1 - input_h, 0)
+                pad_w = max((output_w - 1) * stride_w +
+                            (kernel_w - 1) * self.adap_padding.dilation[1] +
+                            1 - input_w, 0)
+                input_h = input_h + pad_h
+                input_w = input_w + pad_w
+                input_size = (input_h, input_w)
```
Comment (mzr1996) [inline mmdet/models/utils/transformer.py:202]:
```suggestion
            if self.adap_padding:
                pad_h, pad_w = self.adap_padding.get_pad_shape(input_size)
                input_h = input_h + pad_h
                input_w = input_w + pad_w
                input_size = (input_h, input_w)
```

File: mmdet/models/utils/transformer.py [hunk 6]
```diff
@@ -25,6 +29,340 @@
...
+                    (kernel_w - 1) * self.dilation[1] + 1 - input_w, 0)
+        if pad_h > 0 or pad_w > 0:
+            if self.padding == 'corner':
+                x = F.pad(x, [0, pad_w, 0, pad_h])
+            elif self.padding == 'same':
+                x = F.pad(x, [
+                    pad_w // 2, pad_w - pad_w // 2, pad_h // 2,
+                    pad_h - pad_h // 2
+                ])
+        return x
```
Comment (mzr1996) [inline mmdet/models/utils/transformer.py:97]: Maybe we can encapsulate the pad shape calculation as a method.

```suggestion
    def get_pad_shape(self, input_shape):
        input_h, input_w = input_shape
        kernel_h, kernel_w = self.kernel_size
        stride_h, stride_w = self.stride
        output_h = math.ceil(input_h / stride_h)
        output_w = math.ceil(input_w / stride_w)
        pad_h = max((output_h - 1) * stride_h +
                    (kernel_h - 1) * self.dilation[0] + 1 - input_h, 0)
        pad_w = max((output_w - 1) * stride_w +
                    (kernel_w - 1) * self.dilation[1] + 1 - input_w, 0)
        return pad_h, pad_w
        
    def forward(self, x):
        pad_h, pad_w = self.get_pad_shape(x.size()[-2:])
        if pad_h > 0 or pad_w > 0:
            if self.padding == 'corner':
                x = F.pad(x, [0, pad_w, 0, pad_h])
            elif self.padding == 'same':
                x = F.pad(x, [
                    pad_w // 2, pad_w - pad_w // 2, pad_h // 2,
                    pad_h - pad_h // 2
                ])
        return x
```

**Review by mzr1996 — CHANGES_REQUESTED** [review]
A little suggestion.

File: mmdet/models/utils/transformer.py [hunk 7]
```diff
@@ -25,6 +29,340 @@
...
+        >>>     dilation=dilation,
+        >>>     padding="corner")
+        >>> out = adap_pad(input)
+        >>> assert (out.shape[2], out.shape[3]) == (16, 32)
+        >>> input = torch.rand(1, 1, 16, 17)
+        >>> out = adap_pad(input)
+        >>> assert (out.shape[2], out.shape[3]) == (16, 32)
+    """
+
+    def __init__(self, kernel_size=1, stride=1, dilation=1, padding='corner'):
```
Comment (xvjiarui) [inline mmdet/models/utils/transformer.py:63]:
```suggestion
    def __init__(self, kernel_size=1, stride=1, dilation=1, mode='corner'):
```

Comment (xvjiarui) [reply mmdet/models/utils/transformer.py:63]: Mode may be more accurate

Comment (jshilong) [reply mmdet/models/utils/transformer.py:63]: This follows the name of the high version of PyTorch, you can refer to https://pytorch.org/docs/stable/_modules/torch/nn/modules/conv.html#Conv2d

**Review by ZwwWayne — APPROVED** [review]

File: mmdet/models/utils/transformer.py [hunk 8]
```diff
@@ -25,6 +29,332 @@
...
+                stride=stride,
+                dilation=dilation,
+                padding=padding)
+            # disable the padding of conv
+            padding = 0
+        else:
+            self.adap_padding = None
+        padding = to_2tuple(padding)
+
+        self.projection = build_conv_layer(
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:168]: Why do not we use convmodule here but need use build_conv_layer and build_norm_layer?

Comment (jshilong) [reply mmdet/models/utils/transformer.py:168]: because there is a shape related operation between Conv and norm

**Review by ZwwWayne — APPROVED** [review]
